snowdrop-execution/src/Snowdrop/Execution/DbActions/Simple.hs

```diff
@@ -2,134 +2,154 @@
+        -> HChangeSetEl t
+        -> Either CSMappendException (HChangeSetEl t)
+    computeUndo (HMapEl vals) (HChangeSetEl cs) = do
         let processOne m (k, valueop) = case (valueop, k `M.lookup` vals) of
               (New _, Nothing)      -> pure $ M.insert k Rem m
               (Upd _, Just v0)      -> pure $ M.insert k (Upd v0) m
               (NotExisted, Nothing) -> pure $ M.insert k NotExisted m
               (Rem, Just v0)        -> pure $ M.insert k (New v0) m
               _                     -> throwError (CSMappendException k)
-        ChangeSet <$> foldM processOne mempty (M.toList cs)
+        HChangeSetEl <$> foldM processOne mempty (M.toList cs)
 
 chgAccumIter
-    :: (IdSumPrefixed id, Ord id, Applicative m)
-    => (Prefix -> b -> ((id, value) -> b -> b) -> m b)
-    -> SumChangeSet id value
-    -> Prefix
-    -> b
-    -> ((id, value) -> b -> b)
-    -> m b
-chgAccumIter iter accum prefix initB foldF =
-    let newKeysB = M.foldrWithKey (\i v r -> foldF (i, v) r) initB $ getNewKeys accum prefix
-        newFoldF (i, val) b = case queryAccumOne accum i of
-            Nothing         -> foldF (i, val) b
-            Just Rem        -> b
-            Just NotExisted -> b -- something strange happened -- TODO shall we throw error here ?
-            Just (Upd newV) -> foldF (i, newV) b
-            Just (New _)    -> b -- something strange happened
-     in iter prefix newKeysB newFoldF
-
+    :: (Applicative m, RecAll' xs OrdHKey)
+    => DIter' xs m
+    -> DIter (SumChangeSet xs) xs m
+chgAccumIter RNil (SumChangeSet RNil) = pure RNil
+chgAccumIter (iter' :& xs) (unSumCS -> csel' :& ys) =
+    (chgAccumIter' iter' csel' :&) <$> chgAccumIter xs (SumChangeSet ys)
+  where
+    chgAccumIter' :: OrdHKey t => IterAction m t -> HChangeSetEl t -> IterAction m t
+    chgAccumIter' iter ac'@(HChangeSetEl accum) = IterAction $ \initB foldF -> do
+        let newKeysB = M.foldrWithKey (\i v r -> foldF (i, v) r) initB $ csNew ac'
+            newFoldF (i, val) b = case M.lookup i accum of
+                Nothing         -> foldF (i, val) b
+                Just Rem        -> b
+                Just NotExisted -> b -- TODO shall we throw error here ?
+                Just (Upd newV) -> foldF (i, newV) b
+                Just (New _)    -> b -- TODO shall we throw error here?
+        runIterAction iter newKeysB newFoldF
+chgAccumIter _ _ = error "Absurd. Lol, without this I get non exhaustive pattern matching "
 
 chgAccumGetter
-    :: (IdSumPrefixed id, Ord id, Applicative m)
-    => (StateR id -> m (StateP id value))
-    -> SumChangeSet id value
-    -> StateR id
-    -> m (StateP id value)
+    :: (Applicative m, HIntersectable xs xs, Semigroup (HMap xs))
+    => DGetter' xs m
+    -> DGetter (SumChangeSet xs) xs m
 chgAccumGetter getter accum reqIds =
-    bool (unionStateP resp <$> getter reqIds') (pure resp) (null reqIds')
+    bool (unionStateP resp <$> getter reqIds') (pure resp) (rnull reqIds')
```

Comment:
Are you sure `rnull` is what you want here?
My impresion is that you will always get `rnull (hintersect x y)` equal to `False` as instead of checking every component contains empty map you try to check that there is no components.